gk/main.c

```diff
@@ -686,6 +840,549 @@ gk_conf_put(struct gk_config *gk_conf)
+			if (gw_fib == NULL) {
+				ret = -1;
+				goto out;
+			}
+		}
+
+		/* Initialize the fib entry for the Grantor. */
+		ret = convert_str_to_ip(gk_fib->grantor, &gt_addr);
+		if (ret < 0)
+			goto out;
+
+		gt_fib = init_grantor_fib(
+			&gt_addr, gw_fib, gk_conf, &is_new_gt);
+		if (gt_fib == NULL) {
+			ret = -1;
+			goto out;
+		}
+
+		/* Initialize the fib entry for the IP prefix. */
+ 		ip_prefix_len = parse_ip_prefix(
+			gk_fib->ip_prefix, &ip_prefix_addr);
+		if (ip_prefix_len < 0) {
+			ret = -1;
+			goto out;
+		}
+
+		ip_prefix_fib = add_prefix_fib(
+			&ip_prefix_addr, ip_prefix_len, gk_conf);
+		if (ip_prefix_fib == NULL) {
+			ret = -1;
+			goto out;
+		}
+
+		rte_memcpy(ip_prefix_fib, gt_fib, sizeof(*ip_prefix_fib));
+		gt_fib->u.grantor.next_fib->ref_cnt++;
+
+		/*
+	 	 * XXX The nexthop MAC address should be
+	 	 * initialized only after NICs start.
+	 	 */
+		break;
+	}
+
+	case GK_FWD_BACK_NET:
+		/* FALLTHROUGH */
+	case GK_DROP:
+		/* Initialize the fib entry for the IP prefix. */
+ 		ip_prefix_len = parse_ip_prefix(
+			gk_fib->ip_prefix, &ip_prefix_addr);
+		if (ip_prefix_len < 0) {
+			ret = -1;
+			goto out;
+		}
+
+		ip_prefix_fib = add_prefix_fib(
+			&ip_prefix_addr, ip_prefix_len, gk_conf);
+		if (ip_prefix_fib == NULL) {
+			ret = -1;
+			goto out;
+		}
+
+		ip_prefix_fib->action = gk_fib->action;
+
+		break;
+
+	default:
+		RTE_LOG(ERR, GATEKEEPER,
+			"Unknown fib action %u\n",
+			gk_fib->action);
+		ret = -1;
+		goto out;
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gk: add a fib entry [ip prefix = %s, action = %u, grantor = %s, gateway = %s, gateway responsible prefix = %s]\n",
+		gk_fib->ip_prefix, gk_fib->action,
+		gk_fib->grantor, gk_fib->gateway, gk_fib->gw_prefix);
+
+	ret = 0;
+out:
+	return ret;
+}
+
+static int
+lua_init_fib_entries(struct gk_config *gk_conf,
+	struct lua_gk_fib *fib_entries, unsigned int num_fib_entries)
+{
+	unsigned int i;
+	int ret;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	RTE_VERIFY(num_fib_entries <= GK_MAX_NUM_FIB_ENTRIES);
+
+	for (i = 0; i < GK_MAX_NUM_FIB_ENTRIES; i++) {
+		ltbl->fib_tbl[i].action = GK_FIB_MAX;
+		ltbl->fib_tbl6[i].action = GK_FIB_MAX;
+	}
+
+	/* We use the first fib entry with action GK_FWD_NEIGHBOR. */
+	ltbl->fib_tbl[0].action = GK_FWD_NEIGHBOR;
+	ltbl->fib_tbl6[0].action = GK_FWD_NEIGHBOR;
+
+	/* Initialize fib entries. */
+	for (i = 0; i < num_fib_entries; i++) {
+		ret = add_fib_entry(&fib_entries[i], gk_conf);
+		if (ret < 0)
+			goto out;
+	}
+
+	ret = 0;
+
+out:
+	return ret;
+}
+
+int
+lua_init_gk_lpm(
+	struct gk_config *gk_conf, struct net_config *net_conf,
+	struct lua_gk_fib *fib_entries, int num_fib_entries)
+{
+	int ret = 0;
+
+	if (net_conf == NULL || gk_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	gk_conf->net = net_conf;
+
+	if (gk_conf->num_lcores <= 0)
+		goto out;
+
+	/*
+	 * Set up the GK LPM table. We assume that
+	 * all the GK instances are running on the same socket.
+	 */
+	ret = setup_gk_lpm(gk_conf,
```

Comment:
Since what the ``init_lpm()`` does now is only to create C-structs for ``lua_init_gk_lpm()`` and then calls it, we may integrate ``init_lpm()`` to ``run_gk()``. Also, the dynamic configuration should be able to add new fib entries dynamically.